src/gas_metering/mod.rs

```diff
@@ -155,185 +166,243 @@ impl Rules for ConstantCostRules {
-	let gas_meter = backend.gas_meter(&module, rules);
+	let gas_meter = backend.gas_meter(module_info, rules);
 
-	let import_count = module.import_count(elements::ImportCountType::Function) as u32;
-	let functions_space = module.functions_space() as u32;
-	let gas_global_idx = module.globals_space() as u32;
-
-	let mut mbuilder = builder::from_module(module.clone());
+	let import_count = module_info.imported_functions_count;
+	let functions_count = module_info.num_functions();
 
 	// Calculate the indexes and gas function cost,
 	// for external gas function the cost is counted on the host side
-	let (gas_func_idx, total_func, gas_fn_cost) = match gas_meter {
+	let (
+		// Index of the gas function
+		gas_func_idx,
+		// Total number of functions (also the index of the next added function (if added)
+		grow_cnt_func_idx,
+		// Cost of the gas function execution
+		gas_fn_cost,
+	) = match gas_meter {
 		GasMeter::External { module: gas_module, function } => {
 			// Inject the import of the gas function
-			let import_sig = mbuilder
-				.push_signature(builder::signature().with_param(ValueType::I64).build_sig());
-			mbuilder.push_import(
-				builder::import()
-					.module(gas_module)
-					.field(function)
-					.external()
-					.func(import_sig)
-					.build(),
-			);
-
-			(import_count, functions_space + 1, 0)
+			let ty = Type::Func(FuncType::new(vec![ValType::I64], vec![]));
+			module_info.add_import_func(gas_module, function, ty)?;
+
+			(
+				// import_count has become an index of the imported gas function
+				import_count,
+				functions_count + 1,
+				0,
+			)
 		},
-		GasMeter::Internal { global, ref func_instructions, cost } => {
+		GasMeter::Internal { module: _gas_module, global: global_name, ref func, cost } => {
+			let gas_global_idx = module_info.num_globals();
+
 			// Inject the gas counting global
-			mbuilder.push_global(
-				builder::global()
-					.with_type(ValueType::I64)
-					.mutable()
-					.init_expr(Instruction::I64Const(0))
-					.build(),
-			);
-			// Inject the export entry for the gas counting global
-			let ebuilder = builder::ExportBuilder::new();
-			let global_export = ebuilder
-				.field(global)
-				.with_internal(elements::Internal::Global(gas_global_idx))
-				.build();
-			mbuilder.push_export(global_export);
-
-			let func_idx = functions_space;
-
-			// Build local gas function
-			let gas_func_sig =
-				builder::SignatureBuilder::new().with_param(ValueType::I64).build_sig();
-
-			let function = builder::FunctionBuilder::new()
-				.with_signature(gas_func_sig)
-				.body()
-				.with_instructions(func_instructions.clone())
-				.build()
-				.build();
-
-			// Inject local gas function
-			mbuilder.push_function(function);
-
-			(func_idx, func_idx + 1, cost)
+			module_info.add_global(
+				GlobalType { content_type: ValType::I64, mutable: true },
+				&wasm_encoder::ConstExpr::i64_const(0),
+			)?;
+
+			module_info.add_export(global_name, ExportKind::Global, gas_global_idx)?;
```

Comment:
As discussed before, looks like we're not checking the `global_name` existence here, so must be handled by `add_export`?

Reply:
`add_exports()` (new name of `add_export()`) checks now if the export already exists